### ISSUE TITLE
Typechecker allows non-existent fields in embedded objects

### ISSUE THREAD

**Issue by iwahbe** [pulumi/pulumi-yaml#330]
### What happened?

See https://github.com/pulumi/pulumi/issues/10466 for details.

### Steps to reproduce

Run `pulumi up` on 
```yaml
name: yaml-test
runtime: yaml
description: A minimal Pulumi YAML program
resources:
  patch1:
    type: kubernetes:apps/v1:DeploymentPatch
    properties:
      metadata:
        name: "coredns"
        namespace: "kube-system"
        annotations:
          pulumi.com/patchForce: "true"
      spec:
        template:
          containers:
            name: "coredns"
            # Use the version retrieved earlier as the image tag here!
            image: "602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/coredns:v1.8.7-eksbuild.1"
```
With `pulumi-eks = v0.41.2`

### Expected Behavior

We should reject an invalid program.

### Actual Behavior

We let it in.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by aq17** [comment]
@iwahbe @stack72 after incorporating [this PR ](https://github.com/pulumi/pulumi-yaml/pull/320), `pulumi convert` (think it should be that, not `pulumi up` in the description above?) successfully rejected the program 
`error: could not generate output program: :0,0-0: unknown property 'containers' among [spec metadata]; , and 4 other diagnostic(s)`

**Comment by iwahbe** [comment]
@aq17 That error message was introduced by https://github.com/pulumi/pulumi/pull/10654. I do mean `pulumi up`. Pulumi YAML should reject badly typed programs like this one. The error should not be from pulumi convert (which is not designed to provider good error UX), but from the YAML type checker.

**Comment by aq17** [comment]
@iwahbe running `pulumi up` does result in a failed apply
```
     Type                                   Name          Plan     Info
     pulumi:pulumi:Stack                    test-296-dev           
     └─ kubernetes:apps/v1:DeploymentPatch  patch1                 1 error
 
Diagnostics:
  kubernetes:apps/v1:DeploymentPatch (patch1):
    error: patch resources require Server-side Apply mode, which is enabled using the `enableServerSideApply` Provider config
```

Could you clarify the behavior you expect to see?

**Comment by iwahbe** [comment]
@aq17 I expect to see a type-checking error. For example, if you move the `container` property to be top level under the resource, you get the following error:
```yaml
resources:
  patch1:
    type: kubernetes:apps/v1:DeploymentPatch
    properties:
      metadata:
        name: "coredns"
        namespace: "kube-system"
        annotations:
          pulumi.com/patchForce: "true"
      containers:
        name: "coredns"
        # Use the version retrieved earlier as the image tag here!
        image: "602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/coredns:v1.8.7-eksbuild.1"
```
```
  pulumi:pulumi:Stack (yaml-test-dev):
    error: Property containers does not exist on Resource kubernetes:apps/v1:DeploymentPatch
    
      on Pulumi.yaml line 13:
      13:       containers:
      14:         name: "coredns"
      15:         # Use the version retrieved earlier as the image tag here!
      16:         image: "602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/coredns:v1.8.7-eksbuild.1"
    
    Existing properties are: kind, metadata, apiVersion, spec
```

We should get a similar error for nested types.

**Comment by iwahbe** [comment]
Closed by #350
